libplatsupport/src/plat/star64/timer.c

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2023, UNSW
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <inttypes.h>
+#include <platsupport/plat/timer.h>
+#include <stdint.h>
+#include <utils/util.h>
+
+#define MAX_TIMEOUT_NS (STARFIVE_TIMER_MAX_TICKS * (NS_IN_S / STARFIVE_TIMER_TICKS_PER_SECOND))
+
+static void print_regs(starfive_timer_t *timer)
+{
+    printf("   Timer Channel Register State\n");
+    printf("     Interrupt status: 0x%08x @ [%p]\n", timer->regs->status, &timer->regs->status);
+    printf("     Control:          0x%08x @ [%p]\n", timer->regs->ctrl, &timer->regs->ctrl);
+    printf("     Load:             0x%08x @ [%p]\n", timer->regs->load, &timer->regs->load);
+    printf("     Enable:           0x%08x @ [%p]\n", timer->regs->enable, &timer->regs->enable);
+    printf("     Reload:           0x%08x @ [%p]\n", timer->regs->reload, &timer->regs->reload);
+    printf("     Value:            0x%08x @ [%p]\n", timer->regs->value, &timer->regs->value);
+    printf("     Clear Interrupt:  0x%08x @ [%p]\n", timer->regs->intclr, &timer->regs->intclr);
+    printf("     Mask Interrupt:   0x%08x @ [%p]\n", timer->regs->intmask, &timer->regs->intmask);
+}
+
+void starfive_timer_start(starfive_timer_t *timer)
+{
+    assert(timer);
+    assert(timer->regs);
+
+    timer->regs->enable = STARFIVE_TIMER_ENABLED;
+}
+
+void starfive_timer_stop(starfive_timer_t *timer)
+{
+    assert(timer);
+    assert(timer->regs);
+
+    timer->regs->enable = STARFIVE_TIMER_DISABLED;
+}
+
+void starfive_timer_handle_irq(starfive_timer_t *timer)
+{
+    assert(timer);
+    assert(timer->regs);
+
+    timer->value_h += 1;
+
+    while (timer->regs->intclr & STARFIVE_TIMER_INTCLR_BUSY) {
+        /*
+         * Hardware will not currently accept writes to this register.
+         * Wait for this bit to be unset by hardware.
+         */
+    }
+
+    timer->regs->intclr = 1;
+}
+
+uint64_t starfive_timer_get_time(starfive_timer_t *timer)
+{
+    assert(timer);
+    assert(timer->regs);
+
+    /* the timer value counts down from the load value */
+    uint64_t value_l = (uint64_t)(STARFIVE_TIMER_MAX_TICKS - timer->regs->value);
+    uint64_t value_h = (uint64_t)timer->value_h;
+
+    /* Include unhandled interrupt in value_h */
+    if (timer->regs->intclr == 1) {
```

Comment:
How is the register intclr defined, are we really checking the whole register value here, or just the bit 0, because the other bits are reserved? I could not find any document that describes the timer resisters, seem this was never published and this is just taken from some other code?

Reply:
I couldn't find any documentation either other than what was with some code. I did test that this was set as shown when there was an unhandled interrupt. I can add the mask to be defensive.

Reply:
Well, without documentation, it's a well-educated guess that it's following the usual way, where just bit 0 matters. Bitght be safer to add a comment, that it's unclear how exactly this register works and give the source where this code is copied from.

Reply:
To be clear, I didn't copy the source, just used it as documentation.